### PR TITLE
[11.0] add mis_builder_subkpi_template

### PR DESCRIPTION
**Pull request by grindtildeath**
:warning: This is stil a work in progress. :warning: 

Description :

This module extends the functionality of MIS-Builder to add MIS Report Sub-KPI
templates.

Sub-KPI Templates allows to predefine a set of sub-KPIs to be used in a
MIS Report template.

A move_line_domain field is defined on Sub-KPI Template Lines, and allows to
specifiy an extra domain to be applied on the move lines for this sub-KPI.

When entering a KPI expression, it will not be needed to fill the expressions
for each sub-KPI, as the extra move line domain is applied automatically on
KPI expressions when flagging the KPI as multi.

Moreover, this module adds a sub-KPI filter at the MIS Report Instance level to
select which sub-KPIs should be displayed on the report.

I'm on holidays next week, but would be happy to read about any suggestions when I'll be back on Apr 1st.
cc @fclementic2c @yvaucher 

EDIT:  Depends on:
 - [ ] #170

### CODE REVIEW

File: mis_builder_subkpi_template/models/mis_report_instance.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        store=True,
+    )
+    subkpis_filter_active = fields.Boolean(
+        string='Sub-KPIs filter active',
+        default=False,
+    )
+    filter_subkpi_ids = fields.Many2many(
+        'mis.report.subkpi',
+        string='Sub-KPIs to display',
+    )
```
Comment (sbidoul) [inline mis_builder_subkpi_template/models/mis_report_instance.py]: How is this different from the `subkpi_ids` field on `mis.report.instance.period`.
Is it the goal to simplify configuration by filtering subkpis at once for all columns?
In this case, it's best to add this feature in the core module.

Comment (fclementic2c) [reply mis_builder_subkpi_template/models/mis_report_instance.py]: It is the same idea you are right.
In our case, subkpis = department so our accountant want to be able to display a P&L with department in columns then filter quickly from the report to a specific department. 

If we have your blessing, we can add this feature in the core module, yes :)

**Comment by fclementic2c** [comment]
More important for us, @sbidoul : What do you think about the idea of having subkpis tempate that you could define separately from report template.
ex : you create subkpis for departments then you can create several reports based on the template "departments" -> expression will be built automatically with the multi syntax.

**Comment by sbidoul** [comment]
> What do you think about the idea of having subkpis template

I must think about it. This sounds reminiscent of #138. Or a kind of generalization of mis_builder_operating_unit. subkpi template might not be the ideal name for the concept, not sure.

Can you give an example of subkpis template in your case? (screenshot or sample config)

**Comment by fclementic2c** [comment]
@grindtildeath has our test DB, he will send you that next week when he is back

File: mis_builder_subkpi_template/models/mis_report_instance.py (hunk 1 above)
Comment (grindtildeath) [reply mis_builder_subkpi_template/models/mis_report_instance.py]: @sbidoul Yes it is, so I will extract the feature and open another PR to target the core module.

**Comment by grindtildeath** [comment]
@sbidoul 
This is the example I used for our subKPI template departments:
image

We use https://github.com/OCA/account-analytic/tree/11.0/analytic_base_department to have the department_id filled on the analytic account.

It's therefore much easier to define multi subkpis through the onchange, instead of having to repeat the same domain multiple times:
Peek 2019-04-01 15-53

**Comment by sbidoul** [comment]
Thanks for the screenshots. I understand better.

I've the nagging feeling that it's pushing the subkpi concept a bit too far, and it we could reach the same goal more easily with https://github.com/OCA/mis-builder/issues/96. 

But we can let this module incubate in bit. Could you just move the PR to OCA/mis-builder-contrib?

**Comment by grindtildeath** [comment]
@sbidoul 
I extracted all the commits from this PR into multiple PRs :
 - SubKPI file on report instance: https://github.com/OCA/mis-builder/pull/170
 - mis_builder_subkpi_template module: https://github.com/OCA/mis-builder-contrib/pull/1
 - Preparation of subkpi expression: https://github.com/OCA/mis-builder/pull/171

I guess we can close this one and merge the others after reviews ?

**Pull request closed without merging**